api/src/main/java/org/apache/iceberg/types/PruneColumns.java

```diff
@@ -27,10 +27,23 @@
 
 class PruneColumns extends TypeUtil.SchemaVisitor<Type> {
   private final Set<Integer> selected;
+  private final boolean onlySelected;
+
+  /**
+   * Visits a schema and returns only those element's whose id's have been passed as selected
+   * @param selected ids of elements to return
+   * @param onlySelected whether to return only elements which have been selected, no sub elements
```

Comment:
This wording seems a bit unclear to me since the purpose of this class is to return only the selected fields. I think that it misses the context that this affects the fields that are selected when a struct itself is selected.

What about renaming this to `selectFullStructs` or `selectEmptyStructs`? Then the javadoc would be "whether explicitly selected structs should contain all fields or no fields".

Reply:
Sgtm

Reply:
After thinking about this and reading more of the PR, I think that the boolean should be `selectFullTypes` so that it is clear how maps and lists are handled. Additionally, when `selectFullTypes` is `false`, only struct IDs should be supported. Maps and lists should not be explicitly selected or else there is a problem.